app/models/user.rb

```diff
@@ -587,6 +588,29 @@ def unread_notifications
     end
   end
 
+  def all_unread_notifications
+    @all_unread_notifications ||= begin
+      sql = <<~SQL
+        SELECT COUNT(*) FROM (
+          SELECT 1 FROM
+          notifications n
+          LEFT JOIN topics t ON t.id = n.topic_id
+           WHERE t.deleted_at IS NULL AND
+            n.user_id = :user_id AND
+            n.id > :seen_notification_id AND
+            NOT read
+          LIMIT :limit
+        ) AS X
+      SQL
```

Comment:
Can you check that this query is able to use the existing index? I have a feeling we do not have a suitable index here because the existing index has a `high_priority = true` filter which this query is not able to utilise.

Reply:
Executed it on Meta for my user:

```
EXPLAIN ANALYZE SELECT COUNT(*) FROM (
  SELECT 1 FROM
  notifications n
  LEFT JOIN topics t ON t.id = n.topic_id
  WHERE t.deleted_at IS NULL AND
  n.user_id = 16539 AND
  n.id > 14190712 AND
  NOT read
  LIMIT 99
) AS X;
                                                                                 QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=82.57..82.58 rows=1 width=8) (actual time=0.807..0.808 rows=1 loops=1)
   ->  Limit  (cost=74.51..82.55 rows=1 width=4) (actual time=0.805..0.806 rows=0 loops=1)
         ->  Nested Loop Left Join  (cost=74.51..82.55 rows=1 width=4) (actual time=0.804..0.805 rows=0 loops=1)
               Filter: (t.deleted_at IS NULL)
               ->  Bitmap Heap Scan on notifications n  (cost=74.09..78.11 rows=1 width=4) (actual time=0.804..0.804 rows=0 loops=1)
                     Recheck Cond: ((user_id = 16539) AND (NOT read) AND (id > 14190712))
                     ->  BitmapAnd  (cost=74.09..74.09 rows=1 width=0) (actual time=0.800..0.801 rows=0 loops=1)
                           ->  Bitmap Index Scan on idx_notifications_speedup_unread_count  (cost=0.00..25.91 rows=1265 width=0) (actual time=0.433..0.433 rows=41 loops=1)
                                 Index Cond: (user_id = 16539)
                           ->  Bitmap Index Scan on notifications_pkey  (cost=0.00..47.93 rows=1000 width=0) (actual time=0.364..0.364 rows=6421 loops=1)
                                 Index Cond: (id > 14190712)
               ->  Index Only Scan using index_topics_on_id_and_deleted_at on topics t  (cost=0.42..4.44 rows=1 width=12) (never executed)
                     Index Cond: (id = n.topic_id)
                     Heap Fetches: 0
 Planning Time: 1.372 ms
 Execution Time: 0.871 ms
```

Looks reasonable to me, right?

Reply:
Yup let's leave it as it is for now but the query has to do two index scans (idx_notifications_speedup_unread_count and notifications_pkey) instead of one for the `notifications` table. I'm just curious though, can you try adding a temp index for `notifications(user_id, id) WHERE NOT read AND deleted_at IS NULL` and see how the query performs?

Reply:
`deleted_at` is on the `topics` table so I can't include it in the index, but I created this:

```
CREATE INDEX temp_notifications_index ON notifications USING btree (user_id, id) WHERE NOT read;
```

And then `EXPLANE ANALYZE`d the query:

```
discourse=> EXPLAIN ANALYZE SELECT COUNT(*) FROM (
  SELECT 1 FROM
  notifications n
  LEFT JOIN topics t ON t.id = n.topic_id
  WHERE t.deleted_at IS NULL AND
  n.user_id = 16539 AND
  n.id > 14190712 AND
  NOT read
  LIMIT 99
) AS X;
                                                                         QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=25.28..25.29 rows=1 width=8) (actual time=0.022..0.022 rows=1 loops=1)
   ->  Limit  (cost=0.84..25.25 rows=2 width=4) (actual time=0.019..0.020 rows=0 loops=1)
         ->  Nested Loop Left Join  (cost=0.84..25.25 rows=2 width=4) (actual time=0.019..0.019 rows=0 loops=1)
               Filter: (t.deleted_at IS NULL)
               ->  Index Scan using temp_notifications_index on notifications n  (cost=0.42..12.36 rows=2 width=4) (actual time=0.019..0.019 rows=0 loops=1)
                     Index Cond: ((user_id = 16539) AND (id > 14190712))
               ->  Index Only Scan using index_topics_on_id_and_deleted_at on topics t  (cost=0.42..6.44 rows=1 width=12) (never executed)
                     Index Cond: (id = n.topic_id)
                     Heap Fetches: 0
 Planning Time: 1.208 ms
 Execution Time: 0.046 ms
```

And it indeed used the new index as expected.

Reply:
And the execution time is way way faster. Actually let's address this change separately because we currently have 

`index_notifications_read_or_not_high_priority                (user_id,id DESC,read,topic_id) WHERE (read OR (high_priority = false))`

If we can just drop the `WHERE read` from that index, I think it cover more queries.

Reply:
Hmm if we dropped `read` we'd be left with `WHERE high_priority = false`, which would prevent queries that filter by `high_priority = true` from using the index, no?

Reply:
ah yes that is true. I think the following index will be sufficient `(user_id, id, read, high_priority)`